### ISSUE TITLE
AGPS data may be expired. We should try to accelerate fixes by forcing a refresh.

### ISSUE THREAD

**Issue by crankycoder** [mozilla/MozStumbler#1324]
We can help the GPS get a faster time to fix by injecting AGPS assistance data. 

This is dependent on the GPS driver on a per device basis, but this may help in some cases.

The code from https://f-droid.org/repository/browse/?fdfilter=agps&fdid=com.vonglasow.michael.satstat has examples of:
- detecting if AGPS data is expired
- requesting new AGPS data to be injected

**Comment by Djfe** [comment]
I wanted this for so long. I'm using GPS status and toolbox from the play store before every time I go stumbling to get an instant GPS fix.
This would make that step unnecessary.

**Comment by Djfe** [comment]
This could be helpful as well:
https://gitorious.org/android-eeepc/base/source/fd9f42a6967370490cd2c66e628ccbf2eacc4a4b:location/java/com/android/internal/location/GpsLocationProvider.java#L32

**Comment by garvankeeley** [comment]
Injecting AGPS is an impossibility that I am aware of. The GPS interface only accepts SUPL, which is a dedicated connection between the chipset/driver and the SUPL server. I saw the GPS Status app that had a feature to download AGPS data. I don't know how they could do this, as I don't see any injection mechanism in gps.h.

**Comment by maxious** [comment]
There is a generic API in the LocationManager to send extra commands that are chipset provider specific https://developer.android.com/reference/android/location/LocationManager.html#sendExtraCommand(java.lang.String, java.lang.String, android.os.Bundle)

The Qualcomm chipsets at support "force_xtra_injection" and "force_time_injection" (AGPS data and copy system time from CPU)
http://androidforum.tytung.com/android-app-agps-injector-widget-t97.html#p2042
which is present in the app repository mentioned before
https://github.com/mvglasow/satstat/blob/cdae21f4ae1689368817fe2f031f3539ebfa4779/src/com/vonglasow/michael/satstat/GpsEventReceiver.java#L206

**Comment by garvankeeley** [comment]
From reading those links, my understanding: for qualcomm chips, one can force the download from the internet of an AGPS  file 'xtra.bin'. Not clear how that file gets injected to the chip. Any idea?

**Comment by Djfe** [comment]
I don't think that this works on Qualcomm only, because the app GPS Status and Toolbox manages to do it on my Samsung Galaxy Tab 3, too (Intel x86)

maybe you can ask their devs for further advice ;)

**Comment by crankycoder** [comment]
These are two separate sets of APIs.  You can inject time for any chipset.

It's a bit confusing because they're all implemented in https://android.googlesource.com/platform/frameworks/base/+/jb-release/services/jni/com_android_server_location_GpsLocationProvider.cpp

Injecting the XTRA data is only supported by Qualcomm chipsets and it's baked directly into android code.  That API _only_ works on Qualcomm chipsets. 

I found this comment by Mike Lockwood who works on the android source: https://groups.google.com/forum/#!topic/android-platform/3TClTsqtQG0 that the xtra interfaces are specific to qualcomm.
